### PR TITLE
[Kubernetes] Extend plugin to support microk8s

### PR DESCRIPTION
**Pull request by phvalguima**
Currently, we are not supporting microk8s with the
same options as provided by kubernetes. This PR
extends the UbuntuKubernetes to get information
from microk8s.

Following the review, it also extends microk8s plugin
to add the credentials file in the copy spec and
runs the postproc() routine.

Closes: https://github.com/sosreport/sos/issues/3500

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [X ] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [X ] Is the subject and message clear and concise?
- [X ] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [X ] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [X ] Are any related Issues or existing PRs [properly referenced](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) via a Closes (Issue) or Resolved (PR) line?

### CODE REVIEW

File: sos/report/plugins/microk8s.py [hunk 1]
```diff
@@ -28,6 +29,22 @@ class Microk8s(Plugin, UbuntuPlugin):
     packages = ('microk8s',)
 
     microk8s_cmd = "microk8s"
+    kube_cmd = "microk8s kubectl"
+
+    files = (
+        "/var/snap/microk8s/current/credentials/client.config"
```
Comment (arif-ali) [inline sos/report/plugins/microk8s.py:35]: in order for this tuple to work, you need to add a comma at the end

However, this tuple is about triggering the plugin, and not about collecting file config. I would think having the `microk8s` in the `packages` tuple should suffice.

If we need to collect the config file, then we need to add that to the `setup()` function and add

```python
self.add_copy_spec("/var/snap/microk8s/current/credentials/client.config")
```

also noting that we may need to add a `postproc()` function if we need to mask any private information like passwords and private kets from that config file

File: sos/report/plugins/microk8s.py [hunk 2]
```diff
@@ -44,6 +61,7 @@ def setup(self):
         self.add_cmd_output([
             f"microk8s {subcmd}" for subcmd in microk8s_subcmds
         ])
+        super(UbuntuKubernetes, self).setup()
```
Comment (arif-ali) [inline sos/report/plugins/microk8s.py:64]: I've not seen this being done before within `sos` to call out a class outside. I guess if it works, then great. But would be good to get review from @TurboTurtle or @pmoravec on this

**Review by arif-ali — CHANGES_REQUESTED** [review]

File: sos/report/plugins/microk8s.py (hunk 1 above)
Comment (phvalguima) [reply sos/report/plugins/microk8s.py:35]: Hi @arif-ali thaks! Missed the comma indeed :D

re:add_copy_spec
Do I need it? Just asking because we are not doing it in `sos/report/plugins/kubernetes.py::UbuntuKubernetes`

Comment (arif-ali) [reply sos/report/plugins/microk8s.py:35]: The UbuntuKubernetes is there for standard kubernetes gathering.

With this plugin dedicated for `microk8s`, it may be useful to gather that file. Is it useful to gather that file, if so, then we can add it as a `add_copy_spec`, ensuring any obfuscation is done wrt any secrets

Comment (phvalguima) [reply sos/report/plugins/microk8s.py]: Added the copy_spec

File: sos/report/plugins/microk8s.py (hunk 2 above)
Comment (TurboTurtle) [reply sos/report/plugins/microk8s.py]: Hmmmm. Yeah, that's not something we've done and also not something that `Plugin` is designed with the idea of doing.

I think what our typical design would be in this case would be to have a check within the base `kubernetes` class for the installation of microk8s, and if so add the microk8s specific bits there, allowing the rest of the plugin to focus on the shared content between the two. E.G.:

```python
class kubernetes(Plugin)

self.kube_cmd = 'kubectl'

if self.is_installed('microk8s'):
    self.kube_cmd = 'microk8s kubectl'
    self.add_copy_spec(
            "/var/snap/microk8s/current/credentials/client.config"
        )
    microk8s_subcmds = [<whatever isn't shared with kuberentes>]

    self.add_cmd_output([
        f"microk8s {subcmd}" for subcmd in microk8s_subcmds
    ])

[...the rest of the existing kubernetes plugin...]
```

I think this wasn't considered when the `microk8s` plugin was first written due to a significant lack of certain kubernetes bits that we're present in microk8s. If that's changed, I think the above is probably the correct choice (folding microk8s into kubernetes). If that _isn't_ the case, then I'd ask what we gain by calling `super(UbuntuKubernetes, self).setup()` here, since we'd either be getting not-a-lot more collection, or a whole ton of error output, depending on how `microk8s kubectl` behaves when trying to retrieve a resource that doesn't exist in that project.

Comment (arif-ali) [reply sos/report/plugins/microk8s.py]: I think that makes sense, that we just added the `microk8s` bits 

I was already looking at the Kubeflow stuff internally, and actually we can already utilise the kubernetes plugin for his ,and having the microk8s bits within the kubernetes plugin would actually make it easier.

Something on the line of https://paste.ubuntu.com/p/gcrqHkNHxN/ would potentially work I think

Comment (dnegreira) [reply sos/report/plugins/microk8s.py]: Sorry for the late replies :)

I think this all makes sense, the intent of microk8s is to gather the microk8s bits (microk8s status, microk8s config, etc...)
For the kubernetes pod logs we should indeed add the bits in the kubernetes plugin to gather the logs

File: bin/sos [hunk 3]
```diff
@@ -10,6 +10,7 @@
 import sys
 import os
 
+
```
Comment (arif-ali) [inline bin/sos:13]: blank like? can we get this removed, as this is unrelated to the change

File: sos/report/plugins/microk8s.py [hunk 4]
```diff
@@ -48,5 +51,6 @@ def setup(self):
     def postproc(self):
         rsub = r'(certificate-authority-data:|token:)\s.*'
         self.do_cmd_output_sub("microk8s", rsub, r'\1 "**********"')
+        super(Microk8s, self).postproc()
```
Comment (arif-ali) [inline sos/report/plugins/microk8s.py:54]: We don't need this line, I think this was a typo from my diff

File: sos/report/plugins/kubernetes.py [hunk 5]
```diff
@@ -212,6 +212,7 @@ class UbuntuKubernetes(Kubernetes, UbuntuPlugin, DebianPlugin):
     files = (
         '/root/cdk/cdk_addons_kubectl_config',
         '/etc/kubernetes/admin.conf'
```
Comment (arif-ali) [inline sos/report/plugins/kubernetes.py:214]: comma at the end please, I missed that in my diff earlier

**Review by arif-ali — CHANGES_REQUESTED** [review]
a few  comments below.

Then can you squash your commits please.

Thanks for your work on this

File: sos/report/plugins/kubernetes.py [hunk 6]
```diff
@@ -228,6 +229,9 @@ def setup(self):
         for svc in self.services:
             self.add_journal(units=svc)
 
+        if self.is_installed('microk8s'):
+            self.kube_cmd = 'microk8s kubectl'
```
Comment (arif-ali) [inline sos/report/plugins/kubernetes.py:233]: Just testing, and realised the `_kconf` files never got used, and I think, we out to have this if statement before the for loop above. This will ensure that the following command would run

```
microk8s kubectl --kubeconfig=/var/snap/microk8s/current/credentials/client.config
```

Would that make sense? or does `microk8s kubectl` already take care of this?

Comment (phvalguima) [reply sos/report/plugins/kubernetes.py:233]: Hi @arif-ali, `microk8s kubectl` already calls kubectl with admin-level access. I think we do not actually need the client.config at all.

```
$ sudo snap install --classic --channel=latest/edge microk8s
microk8s (edge) v1.29.1 from Canonical✓ installed
$ sudo microk8s kubectl get pods -A
NAMESPACE     NAME                                     READY   STATUS     RESTARTS   AGE
kube-system   calico-node-tmkg7                        0/1     Init:0/2   0          1s
kube-system   calico-kube-controllers-77bd7c5b-98x9p   0/1     Pending    0          1s
kube-system   coredns-864597b5fd-4pq88                 0/1     Pending    0          1s
```

**Review by arif-ali — APPROVED** [review]

**Review by pmoravec — APPROVED** [review]

**Review by TurboTurtle — APPROVED** [review]
